src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,743 @@
+
+    const response = new ServerResponse(incoming);
+    return { incoming, response };
+  }
+
+  listen(...args: unknown[]): this {
+    const [options, callback] = _normalizeArgs(args);
+    if (typeof options.port === 'number' || typeof options.port === 'string') {
+      validatePort(options.port, 'options.port');
+    }
+
+    if (!('port' in options)) {
+      throw new ERR_INVALID_ARG_VALUE(
+        'options',
+        options,
+        'must have the property "port"'
+      );
+    }
+
+    if (this.port != null) {
+      throw new ERR_SERVER_ALREADY_LISTEN();
+    }
+
+    if (portMapper.has(Number(options.port))) {
+      throw new Error(`Port ${options.port} is already in use`);
+    }
+
+    if (callback !== null) {
+      this.once('listening', callback as (...args: unknown[]) => unknown);
+    }
+
+    this.port = Number(options.port);
+    portMapper.set(this.port, { fetch: this.#onRequest.bind(this) });
+    queueMicrotask(() => {
+      callback?.();
+    });
+    return this;
+  }
+
+  getConnections(_cb?: (err: Error | null, count: number) => void): this {
+    // This method is originally implemented in net.Server.
+    // Since we don't implement net.Server yet, we provide this stub implementation for now.
+    // TODO(soon): Revisit this once we implement net.Server
+    return this;
+  }
+
+  ref(): this {
+    // This method is originally implemented in net.Server.
+    // Since we don't implement net.Server yet, we provide this stub implementation for now.
+    // TODO(soon): Revisit this once we implement net.Server
+    return this;
+  }
+
+  unref(): this {
+    // This method is originally implemented in net.Server.
+    // Since we don't implement net.Server yet, we provide this stub implementation for now.
+    // TODO(soon): Revisit this once we implement net.Server
+    return this;
+  }
+}
+
+// We use this handler to not expose this.#fetchResponse to outside world.
+let getServerResponseFetchResponse: (
+  response: ServerResponse
+) => Promise<Response>;
+
+export class ServerResponse<Req extends IncomingMessage = IncomingMessage>
+  extends OutgoingMessage
+  implements _ServerResponse
+{
+  _sent100 = false;
+  _expect_continue = false;
+  override [kOutHeaders]: Record<string, [string, string | string[]]> | null =
+    null;
+
+  statusCode = 200;
+  statusMessage = 'unknown';
+
+  #fetchResponse: Promise<Response>;
+  #encoder = new TextEncoder();
+
+  static {
+    getServerResponseFetchResponse = (
+      response: ServerResponse
+    ): Promise<Response> => {
+      return response.#fetchResponse;
+    };
+  }
+
+  constructor(req: Req, options: ServerOptions = {}) {
+    super(req, options);
+
+    if (req.httpVersionMajor < 1 || req.httpVersionMinor < 1) {
+      this.useChunkedEncodingByDefault = chunkExpression.test(
+        (req.headers.te as string | undefined) ?? ''
+      );
+      this.shouldKeepAlive = false;
+    }
+
+    const { promise, resolve } = Promise.withResolvers<Response>();
```

Comment:
This should grab the reject also and arrange to have it called if the req errors.